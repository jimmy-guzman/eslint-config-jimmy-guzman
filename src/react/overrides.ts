import { EslintRules } from '../models'

export const eslintOverrides: EslintRules = {
  'class-methods-use-this': [
    'error',
    {
      exceptMethods: [
        'render',
        'getInitialState',
        'getDefaultProps',
        'getChildContext',
        'componentWillMount',
        'UNSAFE_componentWillMount',
        'componentDidMount',
        'componentWillReceiveProps',
        'UNSAFE_componentWillReceiveProps',
        'shouldComponentUpdate',
        'componentWillUpdate',
        'UNSAFE_componentWillUpdate',
        'componentDidUpdate',
        'componentWillUnmount',
        'componentDidCatch',
        'getSnapshotBeforeUpdate',
      ],
    },
  ],
}
