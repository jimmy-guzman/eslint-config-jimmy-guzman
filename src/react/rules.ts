import { EslintRules } from '../models'
import { jsxA11yRules } from './jsx-a11y'
import { eslintOverrides } from './overrides'

const hooksRules = <const>{
  'react-hooks/exhaustive-deps': 'error',
  'react-hooks/rules-of-hooks': 'error',
}

export const rules: EslintRules = {
  ...eslintOverrides,
  ...jsxA11yRules,
  ...hooksRules,
  'react/boolean-prop-naming': [
    'error',
    { rule: '^(is|has|should|can|did|will)[A-Z]([A-Za-z0-9]?)+' },
  ],
  /**
   * (react/button-has-type)[https://github.com/yannickcr/eslint-plugin-react/blob/v7.27.0/docs/rules/button-has-type.md]
   * - TYPE: `üö® error`
   * - RATIONALE: turning off due to this issue https://github.com/yannickcr/eslint-plugin-react/issues/1846
   */
  'react/button-has-type': 'off', // revisit
  'react/default-props-match-prop-types': 'error',
  'react/destructuring-assignment': 'off',
  'react/display-name': 'warn',
  'react/forbid-component-props': 'off',
  'react/forbid-dom-props': 'off',
  'react/forbid-elements': 'off',
  'react/forbid-foreign-prop-types': 'off',
  'react/forbid-prop-types': [
    'error',
    {
      forbid: ['any', 'array', 'object'],
      checkContextTypes: true,
      checkChildContextTypes: true,
    },
  ],
  'react/function-component-definition': 'off',
  'react/jsx-boolean-value': ['error', 'never'],
  'react/jsx-curly-brace-presence': 'off',
  'react/jsx-filename-extension': 'off',
  'react/jsx-fragments': ['error', 'syntax'],
  'react/jsx-handler-names': 'off',
  'react/jsx-key': 'warn',
  'react/jsx-max-depth': 'off',
  'react/jsx-no-bind': 'off',
  'react/jsx-no-comment-textnodes': 'warn',
  /**
   * [react/jsx-no-constructed-context-values](https://github.com/yannickcr/eslint-plugin-react/blob/v7.27.0/docs/rules/jsx-no-constructed-context-values.md)
   * - TYPE: `üö® error`
   * - RATIONALE: Prevent needless rerenders and unintended consequences when using object in context
   */
  'react/jsx-no-constructed-context-values': 'error',
  'react/jsx-no-duplicate-props': ['error', { ignoreCase: true }],
  'react/jsx-no-literals': 'off',
  'react/jsx-no-script-url': 'error',
  'react/jsx-no-target-blank': 'warn',
  'react/jsx-no-undef': 'warn',
  'react/jsx-no-useless-fragment': 'error',
  'react/jsx-one-expression-per-line': 'off',
  'react/jsx-pascal-case': 'error',
  'react/jsx-props-no-spreading': 'off',
  'react/jsx-sort-default-props': 'off',
  'react/jsx-sort-props': 'off',
  /**
   * [react/jsx-uses-react](https://github.com/yannickcr/eslint-plugin-react/blob/v7.27.0/docs/rules/jsx-uses-react.md)
   * - TYPE: `üö® error`
   * - RATIONALE: React 17's new jsx transform is favored so this rule is not needed
   */
  'react/jsx-uses-react': 'off',
  'react/jsx-uses-vars': 'warn',
  'react/no-access-state-in-setstate': 'error',
  'react/no-adjacent-inline-elements': 'warn',
  'react/no-array-index-key': 'off',
  /**
   * [react/no-arrow-function-lifecycle](https://github.com/yannickcr/eslint-plugin-react/blob/v7.27.0/docs/rules/no-arrow-function-lifecycle.md)
   * - TYPE: `üö® error`
   * - RATIONALE: It is not necessary to use arrow function for lifecycle methods.
   */
  'react/no-arrow-function-lifecycle': 'error',
  'react/no-children-prop': 'warn',
  'react/no-danger': 'off',
  'react/no-danger-with-children': 'warn',
  'react/no-deprecated': 'warn',
  'react/no-did-mount-set-state': 'error',
  'react/no-did-update-set-state': 'error',
  'react/no-direct-mutation-state': 'warn',
  'react/no-find-dom-node': 'warn',
  /**
   * [react/no-invalid-html-attribute](https://github.com/yannickcr/eslint-plugin-react/blob/v7.27.0/docs/rules/no-invalid-html-attribute.md)
   * - TYPE: `‚ôªÔ∏è best-practices`
   * - RATIONALE: Minimize confusion by using wrong values for `rel` html attribute
   */
  'react/no-invalid-html-attribute': 'error',
  'react/no-is-mounted': 'warn',
  'react/no-multi-comp': 'off',
  /**
   * [react/no-namespace](https://github.com/yannickcr/eslint-plugin-react/blob/v7.27.0/docs/rules/no-namespace.md)
   * - TYPE: `üö® error`
   * - RATIONALE: avoids namespaces in React elements since they're not supported
   */
  'react/no-namespace': 'error',
  'react/no-redundant-should-component-update': 'error',
  'react/no-render-return-value': 'warn',
  'react/no-set-state': 'off',
  'react/no-string-refs': 'warn',
  'react/no-this-in-sfc': 'error',
  'react/no-typos': 'error',
  'react/no-unescaped-entities': 'warn',
  'react/no-unknown-property': 'warn',
  'react/no-unsafe': 'off',
  /**
   * [react/no-unstable-nested-components](https://github.com/yannickcr/eslint-plugin-react/blob/v7.27.0/docs/rules/no-unstable-nested-components.md)
   * - TYPE: `üö® error`
   * - RATIONALE: Prevents unstable components by avoiding creating components inside components without memoization
   */
  'react/no-unstable-nested-components': 'error',
  /**
   * [react/no-unused-class-component-methods](https://github.com/yannickcr/eslint-plugin-react/blob/v7.27.0/docs/rules/no-unused-class-component-methods.md)
   * - TYPE: `üö® error`
   * - RATIONALE: unused methods can take up space in the code and can lead to confusion by readers
   */
  'react/no-unused-class-component-methods': 'error',
  'react/no-unused-prop-types': 'error',
  'react/no-unused-state': 'error',
  'react/no-will-update-set-state': 'error',
  'react/prefer-es6-class': 'off',
  /**
   * [react/prefer-exact-props](https://github.com/yannickcr/eslint-plugin-react/blob/v7.27.0/docs/rules/prefer-exact-props.md)
   * - TYPE: `üö® error`
   * - RATIONALE:
   */
  'react/prefer-exact-props': 'off', // revisit
  'react/prefer-read-only-props': 'off',
  'react/prefer-stateless-function': 'off',
  'react/prop-types': 'error',
  /**
   * [react/react-in-jsx-scope](https://github.com/yannickcr/eslint-plugin-react/blob/v7.27.0/docs/rules/react-in-jsx-scope.md)
   * - TYPE: `üö® error`
   * - RATIONALE: React 17's new jsx transform is favored so this rule is not needed
   */
  'react/react-in-jsx-scope': 'off',
  'react/require-default-props': 'off',
  'react/require-optimization': 'off',
  'react/require-render-return': 'warn',
  'react/self-closing-comp': 'error',
  'react/sort-comp': 'off',
  'react/sort-prop-types': 'off',
  'react/state-in-constructor': 'off',
  'react/static-property-placement': 'off',
  'react/style-prop-object': 'error',
  'react/void-dom-elements-no-children': 'error',
}
