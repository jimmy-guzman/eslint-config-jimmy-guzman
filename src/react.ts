import { rules } from './rules/react'

export = {
  extends: ['prettier'],
  plugins: ['react', 'react-hooks', 'jsx-a11y'],
  rules,
  settings: { react: { version: 'detect' } },
}
