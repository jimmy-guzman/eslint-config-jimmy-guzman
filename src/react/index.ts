import { rules } from './rules'

export = {
  plugins: ['react', 'react-hooks', 'jsx-a11y'],
  rules,
  settings: { react: { version: 'detect' } },
}
