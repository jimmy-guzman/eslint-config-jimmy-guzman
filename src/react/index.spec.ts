import react from './index'

describe('react', () => {
  it('should export react rules config', () => {
    expect(Object.keys(react.rules)).toHaveLength(131)
    expect(react).toMatchSnapshot()
  })
})
