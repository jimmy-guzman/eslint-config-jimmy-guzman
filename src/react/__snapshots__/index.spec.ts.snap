// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`react > should export react rules config 1`] = `
{
  "extends": [
    "prettier",
  ],
  "plugins": [
    "react",
    "react-hooks",
    "jsx-a11y",
  ],
  "rules": {
    "class-methods-use-this": [
      "error",
      {
        "exceptMethods": [
          "render",
          "getInitialState",
          "getDefaultProps",
          "getChildContext",
          "componentWillMount",
          "UNSAFE_componentWillMount",
          "componentDidMount",
          "componentWillReceiveProps",
          "UNSAFE_componentWillReceiveProps",
          "shouldComponentUpdate",
          "componentWillUpdate",
          "UNSAFE_componentWillUpdate",
          "componentDidUpdate",
          "componentWillUnmount",
          "componentDidCatch",
          "getSnapshotBeforeUpdate",
        ],
      },
    ],
    "jsx-a11y/alt-text": "error",
    "jsx-a11y/anchor-ambiguous-text": "off",
    "jsx-a11y/anchor-has-content": "error",
    "jsx-a11y/anchor-is-valid": "error",
    "jsx-a11y/aria-activedescendant-has-tabindex": "error",
    "jsx-a11y/aria-props": "error",
    "jsx-a11y/aria-proptypes": "error",
    "jsx-a11y/aria-role": "error",
    "jsx-a11y/aria-unsupported-elements": "error",
    "jsx-a11y/autocomplete-valid": "error",
    "jsx-a11y/click-events-have-key-events": "error",
    "jsx-a11y/control-has-associated-label": [
      "off",
      {
        "ignoreElements": [
          "audio",
          "canvas",
          "embed",
          "input",
          "textarea",
          "tr",
          "video",
        ],
        "ignoreRoles": [
          "grid",
          "listbox",
          "menu",
          "menubar",
          "radiogroup",
          "row",
          "tablist",
          "toolbar",
          "tree",
          "treegrid",
        ],
        "includeRoles": [
          "alert",
          "dialog",
        ],
      },
    ],
    "jsx-a11y/heading-has-content": "error",
    "jsx-a11y/html-has-lang": "error",
    "jsx-a11y/iframe-has-title": "error",
    "jsx-a11y/img-redundant-alt": "error",
    "jsx-a11y/interactive-supports-focus": [
      "error",
      {
        "tabbable": [
          "button",
          "checkbox",
          "link",
          "progressbar",
          "searchbox",
          "slider",
          "spinbutton",
          "switch",
          "textbox",
        ],
      },
    ],
    "jsx-a11y/label-has-associated-control": "error",
    "jsx-a11y/label-has-for": "off",
    "jsx-a11y/lang": "error",
    "jsx-a11y/media-has-caption": "error",
    "jsx-a11y/mouse-events-have-key-events": "error",
    "jsx-a11y/no-access-key": "error",
    "jsx-a11y/no-aria-hidden-on-focusable": "off",
    "jsx-a11y/no-autofocus": "error",
    "jsx-a11y/no-distracting-elements": "error",
    "jsx-a11y/no-interactive-element-to-noninteractive-role": "error",
    "jsx-a11y/no-noninteractive-element-interactions": [
      "error",
      {
        "body": [
          "onError",
          "onLoad",
        ],
        "iframe": [
          "onError",
          "onLoad",
        ],
        "img": [
          "onError",
          "onLoad",
        ],
      },
    ],
    "jsx-a11y/no-noninteractive-element-to-interactive-role": "error",
    "jsx-a11y/no-noninteractive-tabindex": "error",
    "jsx-a11y/no-redundant-roles": "error",
    "jsx-a11y/no-static-element-interactions": "error",
    "jsx-a11y/prefer-tag-over-role": "off",
    "jsx-a11y/role-has-required-aria-props": "error",
    "jsx-a11y/role-supports-aria-props": "error",
    "jsx-a11y/scope": "error",
    "jsx-a11y/tabindex-no-positive": "error",
    "react-hooks/exhaustive-deps": "error",
    "react-hooks/rules-of-hooks": "error",
    "react/boolean-prop-naming": [
      "error",
      {
        "rule": "^(is|has|should|can|did|will)[A-Z]([A-Za-z0-9]?)+",
      },
    ],
    "react/button-has-type": "off",
    "react/default-props-match-prop-types": "error",
    "react/destructuring-assignment": "off",
    "react/display-name": "warn",
    "react/forbid-component-props": "off",
    "react/forbid-dom-props": "off",
    "react/forbid-elements": "off",
    "react/forbid-foreign-prop-types": "off",
    "react/forbid-prop-types": [
      "error",
      {
        "checkChildContextTypes": true,
        "checkContextTypes": true,
        "forbid": [
          "any",
          "array",
          "object",
        ],
      },
    ],
    "react/function-component-definition": "off",
    "react/jsx-boolean-value": [
      "error",
      "never",
    ],
    "react/jsx-curly-brace-presence": "off",
    "react/jsx-filename-extension": "off",
    "react/jsx-fragments": [
      "error",
      "syntax",
    ],
    "react/jsx-handler-names": "off",
    "react/jsx-key": "warn",
    "react/jsx-max-depth": "off",
    "react/jsx-no-bind": "off",
    "react/jsx-no-comment-textnodes": "warn",
    "react/jsx-no-constructed-context-values": "error",
    "react/jsx-no-duplicate-props": [
      "error",
      {
        "ignoreCase": true,
      },
    ],
    "react/jsx-no-literals": "off",
    "react/jsx-no-script-url": "error",
    "react/jsx-no-target-blank": "warn",
    "react/jsx-no-undef": "warn",
    "react/jsx-no-useless-fragment": "error",
    "react/jsx-one-expression-per-line": "off",
    "react/jsx-pascal-case": "error",
    "react/jsx-props-no-spreading": "off",
    "react/jsx-sort-default-props": "off",
    "react/jsx-sort-props": "off",
    "react/jsx-uses-react": "off",
    "react/jsx-uses-vars": "warn",
    "react/no-access-state-in-setstate": "error",
    "react/no-adjacent-inline-elements": "warn",
    "react/no-array-index-key": "off",
    "react/no-arrow-function-lifecycle": "error",
    "react/no-children-prop": "warn",
    "react/no-danger": "off",
    "react/no-danger-with-children": "warn",
    "react/no-deprecated": "warn",
    "react/no-did-mount-set-state": "error",
    "react/no-did-update-set-state": "error",
    "react/no-direct-mutation-state": "warn",
    "react/no-find-dom-node": "warn",
    "react/no-invalid-html-attribute": "error",
    "react/no-is-mounted": "warn",
    "react/no-multi-comp": "off",
    "react/no-namespace": "error",
    "react/no-redundant-should-component-update": "error",
    "react/no-render-return-value": "warn",
    "react/no-set-state": "off",
    "react/no-string-refs": "warn",
    "react/no-this-in-sfc": "error",
    "react/no-typos": "error",
    "react/no-unescaped-entities": "warn",
    "react/no-unknown-property": "warn",
    "react/no-unsafe": "off",
    "react/no-unstable-nested-components": "error",
    "react/no-unused-class-component-methods": "error",
    "react/no-unused-prop-types": "error",
    "react/no-unused-state": "error",
    "react/no-will-update-set-state": "error",
    "react/prefer-es6-class": "off",
    "react/prefer-exact-props": "off",
    "react/prefer-read-only-props": "off",
    "react/prefer-stateless-function": "off",
    "react/prop-types": "error",
    "react/react-in-jsx-scope": "off",
    "react/require-default-props": "off",
    "react/require-optimization": "off",
    "react/require-render-return": "warn",
    "react/self-closing-comp": "error",
    "react/sort-comp": "off",
    "react/sort-prop-types": "off",
    "react/state-in-constructor": "off",
    "react/static-property-placement": "off",
    "react/style-prop-object": "error",
    "react/void-dom-elements-no-children": "error",
  },
  "settings": {
    "react": {
      "version": "detect",
    },
  },
}
`;
